specification/securitycenter/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json

```diff
@@ -0,0 +1,1975 @@
+                    }
+                }
+            }
+        },
+        "JitNetworkAccessPolicy": {
+            "type": "object",
+            "required": [
+                "properties"
+            ],
+            "properties": {
+                "properties": {
+                    "required": [
+                        "virtualMachines"
+                    ],
+                    "type": "object",
+                    "properties": {
+                        "virtualMachines": {
+                            "type": "array",
+                            "description": "Configurations for Microsoft.Compute/virtualMachine resource type.",
+                            "items": {
+                                "$ref": "#/definitions/JitNetworkAccessRequest"
+                            }
+                        },
+                        "requests": {
+                            "type": "array",
+                            "items": {
+                                "$ref": "#/definitions/JitNetworkAccessPolicyVirtualMachine"
+                            }
+                        },
+                        "provisioningState": {
+                            "type": "string",
+                            "readOnly": true,
+                            "description": "Gets the provisioning state of the Just-in-Time policy.",
+                            "enum": [
+                                "Succeeded",
+                                "Failed"
+                            ],
+                            "x-ms-enum": {
+                                "name": "provisioningState",
+                                "modelAsString": false
+                            }
+                        }
+                    }
+                }
+            },
+            "allOf": [{
+                    "$ref": "#/definitions/Resource"
+                },
+                {
+                    "$ref": "#/definitions/Kind"
+                },
+                {
+                    "$ref": "#/definitions/Location"
+                }
+            ]
+        },
+        "JitNetworkAccessPolicyVirtualMachine": {
+            "type": "object",
+            "required": [
+                "id",
+                "ports"
+            ],
+            "properties": {
+                "id": {
+                    "type": "string",
+                    "description": "Resource ID of the virtual machine that is linked to this policy"
+                },
+                "ports": {
+                    "type": "array",
+                    "description": "Port configurations for the virtual machine",
+                    "items": {
+                        "required": [
+                            "allowedSourceAddressPrefix",
+                            "maxRequestAccessDuration",
+                            "number",
+                            "protocol"
+                        ],
+                        "type": "object",

```

Comment:
simplified the complex objects
let us know if there are still objects that are considered complex by you

Reply:
@chlahav - They are all over the place in both the swagger specs in this PR. 
> It will be time consuming for me to scan each and every line as I have other tasks to accomplish.

However I can provide you a pattern to look for and I would expect you to make sure that you do a good job at it (since you are the spec author).

**Pattern to look for:**
In a swagger spec, inside the top level`"definitions": {}` object model definitions are defined. The definitions object is like a **dictionary**, where the **key is a model name** an the **value is a complex object that defines it's type, properties and other important things**. Now, when you define that object (the value of a model definition) there are some important things that you need to take note of 
- You must provide the `"type": "..."` of that object
- any `"required": [ . . . ]` properties of that object (**NOTE**: This should be provided only for the properties of that object and **NOT** the properties of the object in the `allOf` array. The convention is that when you define that object (for which you provided a reference in the `allOf` array) you would define it's required properties in its definition)
- You define the "properties" of that object. While describing those properties, it is perfectly fine to define schema of properties **inline if they are primitive types** ("string", "number", "integer", "boolean", OR collection (array, dictionary) of primitives). However, **if the property is a complex object** then the convention is to describe that property as a reference and then define the schema/definition of that complex type as a top level model definition in the `"definitions": {}` object in the swagger spec.

The above convention applies to **model definitions**, schema for **request body** parameters and  schema for **response body** of a response status code.

Hope this makes things clear. Let me know if you have more questions. 

I am looking forward to you making changes in both the swagger specs based on the above explanation and @sergey-shandar making sure that the convention is followed.

Reply:
thanks for the clarification
I hope that its better now


---
In reply to: [180822563](https://github.com/Azure/azure-rest-api-specs/pull/2753#discussion_r180822563) [](ancestors = 180822563)